cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/worker/system/SystemWorkerServiceLauncher.java

```diff
@@ -0,0 +1,182 @@
+import org.apache.hadoop.conf.Configuration;
+import org.apache.twill.api.ResourceSpecification;
+import org.apache.twill.api.TwillController;
+import org.apache.twill.api.TwillPreparer;
+import org.apache.twill.api.TwillRunner;
+import org.apache.twill.common.Threads;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.File;
+import java.io.IOException;
+import java.io.Writer;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+
+public class SystemWorkerServiceLauncher extends AbstractScheduledService {
+  private static final Logger LOG = LoggerFactory.getLogger(SystemWorkerServiceLauncher.class);
+  private final CConfiguration cConf;
+  private final Configuration hConf;
+
+  private final TwillRunner twillRunner;
+  private TwillController twillController;
+
+  private ScheduledExecutorService executor;
+
+  @Inject
+  public SystemWorkerServiceLauncher(CConfiguration cConf, Configuration hConf,
+      TwillRunner twillRunner) {
+    this.cConf = cConf;
+    this.hConf = hConf;
+    this.twillRunner = twillRunner;
+  }
+
+  @Override
+  protected void startUp() throws Exception {
+    LOG.info("Starting SystemServiceLauncher.");
+  }
+
+  @Override
+  protected void shutDown() throws Exception {
+    LOG.info("Shutting down SystemServiceLauncher.");
+    try {
+      if (twillController != null) {
+        twillController.terminate().get(10, TimeUnit.SECONDS);
+      }
+    } catch (Exception e) {
+      LOG.warn("Failed to terminate SystemServiceLauncher run", e);
+    }
+    if (executor != null) {
+      executor.shutdownNow();
+    }
+    LOG.info("Shutting down SystemServiceLauncher has completed.");
+  }
+
+  @Override
+  protected void runOneIteration() throws Exception {
+    run();
+  }
+
+  @Override
+  protected Scheduler scheduler() {
+    return Scheduler.newFixedRateSchedule(0,
+        cConf.getInt(Constants.TaskWorker.POOL_CHECK_INTERVAL), TimeUnit.SECONDS);
+  }
+
+  public void run() {
+    TwillController activeController = null;
+    for (TwillController controller : twillRunner.lookup(SystemWorkerTwillApplication.NAME)) {
+      // If detected more than one controller, terminate those extra controllers.
+      if (activeController != null) {
+        controller.terminate();
+      } else {
+        activeController = controller;
+      }
+    }
+    // If there is no system worker runner running, create one
+    if (activeController == null) {
+      try {
+        Path tmpDir = new File(cConf.get(Constants.CFG_LOCAL_DATA_DIR),
+            cConf.get(Constants.AppFabric.TEMP_DIR)).toPath();
+        Files.createDirectories(tmpDir);
+
+        Path runDir = Files.createTempDirectory(tmpDir, "system.worker.launcher");
+        try {
+          // Unset the internal certificate path since certificate is stored cdap-security which
```

Comment:
This applies to worker pod running user code, but not system pod

Reply:
Thanks for pointing out. Removed code for unset.